File: pulsedev/docker/code_executor/executor.js
// JavaScript code executor for PulseDev
// Listens for code execution commands and runs them in the container

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Configuration
const WORKSPACE_DIR = process.env.WORKSPACE_DIR || '/home/pulsedev/workspace';
const MAX_OUTPUT_SIZE = process.env.MAX_OUTPUT_SIZE || 1024 * 1024; // 1MB
const DEFAULT_TIMEOUT = process.env.TIMEOUT || 30; // 30 seconds

// Ensure workspace directory exists
if (!fs.existsSync(WORKSPACE_DIR)) {
    fs.mkdirSync(WORKSPACE_DIR, { recursive: true });
}

// Helper function to generate a safe temporary filename
function getTempFilename(extension) {
    const hash = crypto.randomBytes(8).toString('hex');
    return path.join('/tmp', `pulsedev_${hash}.${extension}`);
}

// Execute code and return result
function executeCode(language, code, timeout = DEFAULT_TIMEOUT) {
    console.log(`Executing ${language} code (timeout: ${timeout}s)`);
    
    let extension, command;
    switch (language.toLowerCase()) {
        case 'javascript':
        case 'js':
            extension = 'js';
            command = 'node';
            break;
        case 'typescript':
        case 'ts':
            extension = 'ts';
            command = 'npx ts-node';
            break;
        case 'python':
        case 'py':
            extension = 'py';
            command = 'python3';
            break;
        case 'shell':
        case 'bash':
        case 'sh':
            extension = 'sh';
            command = 'bash';
            break;
        default:
            return {
                success: false,
                error: `Unsupported language: ${language}`,
                output: '',
                executionTime: 0
            };
    }
    
    // Write code to temp file
    const tempFile = getTempFilename(extension);
    fs.writeFileSync(tempFile, code);
    
    // Make shell scripts executable
    if (extension === 'sh') {
        fs.chmodSync(tempFile, '755');
    }
    
    // Execute with timeout
    const startTime = Date.now();
    let result = {
        success: false,
        error: null,
        output: '',
        executionTime: 0
    };
    
    try {
        // Execute with timeout using 'timeout' command
        const output = execSync(`timeout ${timeout}s ${command} ${tempFile}`, {
            encoding: 'utf8',
            maxBuffer: MAX_OUTPUT_SIZE
        });
        
        result.success = true;
        result.output = output;
    } catch (err) {
        result.success = false;
        
        if (err.status === 124) {
            result.error = `Execution timed out after ${timeout} seconds`;
        } else {
            result.error = err.message;
            result.output = err.stdout || '';
        }
    } finally {
        // Calculate execution time
        result.executionTime = (Date.now() - startTime) / 1000;
        
        // Clean up temp file
        try {
            fs.unlinkSync(tempFile);
        } catch (err) {
            console.error(`Failed to remove temp file ${tempFile}: ${err.message}`);
        }
    }
    
    return result;
}

// Execute file from workspace
function executeFile(filePath, timeout = DEFAULT_TIMEOUT) {
    console.log(`Executing file: ${filePath} (timeout: ${timeout}s)`);
    
    // Resolve path
    const fullPath = path.resolve(WORKSPACE_DIR, filePath);
    
    // Verify path is within workspace for security
    if (!fullPath.startsWith(WORKSPACE_DIR)) {
        return {
            success: false,
            error: 'Security error: Attempted to access file outside workspace',
            output: '',
            executionTime: 0
        };
    }
    
    // Check if file exists
    if (!fs.existsSync(fullPath)) {
        return {
            success: false,
            error: `File not found: ${filePath}`,
            output: '',
            executionTime: 0
        };
    }
    
    // Determine language from extension
    const extension = path.extname(fullPath).toLowerCase().slice(1);
    let command;
    
    switch (extension) {
        case 'js':
            command = 'node';
            break;
        case 'ts':
            command = 'npx ts-node';
            break;
        case 'py':
            command = 'python3';
            break;
        case 'sh':
            command = 'bash';
            break;
        default:
            return {
                success: false,
                error: `Unsupported file type: ${extension}`,
                output: '',
                executionTime: 0
            };
    }
    
    // Execute with timeout
    const startTime = Date.now();
    let result = {
        success: false,
        error: null,
        output: '',
        executionTime: 0,
        fileName: filePath
    };
    
    try {
        // Execute with timeout using 'timeout' command
        const output = execSync(`timeout ${timeout}s ${command} ${fullPath}`, {
            encoding: 'utf8',
            maxBuffer: MAX_OUTPUT_SIZE
        });
        
        result.success = true;
        result.output = output;
    } catch (err) {
        result.success = false;
        
        if (err.status === 124) {
            result.error = `Execution timed out after ${timeout} seconds`;
        } else {
            result.error = err.message;
            result.output = err.stdout || '';
        }
    } finally {
        // Calculate execution time
        result.executionTime = (Date.now() - startTime) / 1000;
    }
    
    return result;
}

// Handle command line arguments
const args = process.argv.slice(2);

if (args.length >= 2) {
    const command = args[0];
    
    if (command === 'exec-code') {
        const language = args[1];
        const code = fs.readFileSync(0, 'utf8'); // Read from stdin
        const timeout = args[2] ? parseInt(args[2], 10) : DEFAULT_TIMEOUT;
        
        const result = executeCode(language, code, timeout);
        console.log(JSON.stringify(result, null, 2));
    } else if (command === 'exec-file') {
        const filePath = args[1];
        const timeout = args[2] ? parseInt(args[2], 10) : DEFAULT_TIMEOUT;
        
        const result = executeFile(filePath, timeout);
        console.log(JSON.stringify(result, null, 2));
    } else {
        console.error(`Unknown command: ${command}`);
        process.exit(1);
    }
} else {
    console.log('JavaScript executor for PulseDev');
    console.log('Usage:');
    console.log('  node executor.js exec-code <language> [timeout] < code.js');
    console.log('  node executor.js exec-file <filepath> [timeout]');
}

File: pulsedev/docker/code_executor/executor.py
#!/usr/bin/env python3
"""
Python code executor for PulseDev
Provides a standardized interface for executing code in various languages
"""

import os
import sys
import json
import time
import tempfile
import subprocess
import signal
import uuid
import traceback
from pathlib import Path

# Configuration
WORKSPACE_DIR = os.environ.get('WORKSPACE_DIR', '/home/pulsedev/workspace')
MAX_OUTPUT_SIZE = int(os.environ.get('MAX_OUTPUT_SIZE', 1024 * 1024))  # 1MB
DEFAULT_TIMEOUT = int(os.environ.get('TIMEOUT', 30))  # 30 seconds

# Ensure workspace directory exists
Path(WORKSPACE_DIR).mkdir(parents=True, exist_ok=True)

def execute_code(language, code, timeout=DEFAULT_TIMEOUT):
    """Execute code in the specified language and return the result"""
    print(f"Executing {language} code (timeout: {timeout}s)")
    
    # Map language to command and file extension
    language_map = {
        'python': ('python3', 'py'),
        'py': ('python3', 'py'),
        'javascript': ('node', 'js'),
        'js': ('node', 'js'),
        'typescript': ('npx ts-node', 'ts'),
        'ts': ('npx ts-node', 'ts'),
        'bash': ('bash', 'sh'),
        'sh': ('bash', 'sh'),
    }
    
    if language.lower() not in language_map:
        return {
            'success': False,
            'error': f"Unsupported language: {language}",
            'output': '',
            'execution_time': 0
        }
    
    command, extension = language_map[language.lower()]
    
    # Create temporary file
    with tempfile.NamedTemporaryFile(suffix=f'.{extension}', delete=False) as temp_file:
        temp_path = temp_file.name
        temp_file.write(code.encode('utf-8'))
    
    # Make shell scripts executable
    if extension == 'sh':
        os.chmod(temp_path, 0o755)
    
    # Execute with timeout
    start_time = time.time()
    result = {
        'success': False,
        'error': None,
        'output': '',
        'execution_time': 0
    }
    
    try:
        # Execute with timeout
        process = subprocess.Popen(
            f"timeout {timeout}s {command} {temp_path}",
            shell=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True
        )
        
        output, _ = process.communicate()
        
        if process.returncode == 124:  # timeout exit code
            result['success'] = False
            result['error'] = f"Execution timed out after {timeout} seconds"
        elif process.returncode != 0:
            result['success'] = False
            result['error'] = f"Execution failed with exit code {process.returncode}"
            result['output'] = output
        else:
            result['success'] = True
            result['output'] = output
    
    except Exception as e:
        result['success'] = False
        result['error'] = str(e)
        result['output'] = traceback.format_exc()
    
    finally:
        # Calculate execution time
        result['execution_time'] = time.time() - start_time
        
        # Clean up temp file
        try:
            os.unlink(temp_path)
        except Exception as e:
            print(f"Failed to remove temp file {temp_path}: {e}")
    
    return result

def execute_file(file_path, timeout=DEFAULT_TIMEOUT):
    """Execute a file from the workspace"""
    print(f"Executing file: {file_path} (timeout: {timeout}s)")
    
    # Resolve path
    full_path = os.path.abspath(os.path.join(WORKSPACE_DIR, file_path))
    
    # Verify path is within workspace for security
    if not full_path.startswith(WORKSPACE_DIR):
        return {
            'success': False,
            'error': 'Security error: Attempted to access file outside workspace',
            'output': '',
            'execution_time': 0,
            'file_name': file_path
        }
    
    # Check if file exists
    if not os.path.exists(full_path):
        return {
            'success': False,
            'error': f"File not found: {file_path}",
            'output': '',
            'execution_time': 0,
            'file_name': file_path
        }
    
    # Determine language from extension
    extension = os.path.splitext(full_path)[1].lower()[1:]
    
    command_map = {
        'py': 'python3',
        'js': 'node',
        'ts': 'npx ts-node',
        'sh': 'bash',
    }
    
    if extension not in command_map:
        return {
            'success': False,
            'error': f"Unsupported file type: {extension}",
            'output': '',
            'execution_time': 0,
            'file_name': file_path
        }
    
    command = command_map[extension]
    
    # Execute with timeout
    start_time = time.time()
    result = {
        'success': False,
        'error': None,
        'output': '',
        'execution_time': 0,
        'file_name': file_path
    }
    
    try:
        # Execute with timeout
        process = subprocess.Popen(
            f"timeout {timeout}s {command} {full_path}",
            shell=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True
        )
        
        output, _ = process.communicate()
        
        if process.returncode == 124:  # timeout exit code
            result['success'] = False
            result['error'] = f"Execution timed out after {timeout} seconds"
        elif process.returncode != 0:
            result['success'] = False
            result['error'] = f"Execution failed with exit code {process.returncode}"
            result['output'] = output
        else:
            result['success'] = True
            result['output'] = output
    
    except Exception as e:
        result['success'] = False
        result['error'] = str(e)
        result['output'] = traceback.format_exc()
    
    finally:
        # Calculate execution time
        result['execution_time'] = time.time() - start_time
    
    return result

def main():
    """Main entry point"""
    if len(sys.argv) >= 3:
        command = sys.argv[1]
        
        if command == 'exec-code':
            language = sys.argv[2]
            code = sys.stdin.read()  # Read from stdin
            timeout = int(sys.argv[3]) if len(sys.argv) > 3 else DEFAULT_TIMEOUT
            
            result = execute_code(language, code, timeout)
            print(json.dumps(result, indent=2))
        
        elif command == 'exec-file':
            file_path = sys.argv[2]
            timeout = int(sys.argv[3]) if len(sys.argv) > 3 else DEFAULT_TIMEOUT
            
            result = execute_file(file_path, timeout)
            print(json.dumps(result, indent=2))
        
        else:
            print(f"Unknown command: {command}")
            sys.exit(1)
    
    else:
        print("Python executor for PulseDev")
        print("Usage:")
        print("  python executor.py exec-code <language> [timeout] < code.py")
        print("  python executor.py exec-file <filepath> [timeout]")

if __name__ == "__main__":
    main()